srcpkgs/python3-pplpy/template

```diff
@@ -1,15 +1,20 @@
 # Template file for 'python3-pplpy'
 pkgname=python3-pplpy
 version=0.8.7
-revision=2
+revision=3
 build_style=python3-module
 hostmakedepends="python3-setuptools python3-Cython"
 makedepends="python3-cysignals python3-gmpy2 python3-devel gmp-devel
  gmpxx-devel ppl-devel mpfr-devel libmpc-devel pari-devel"
+depends="python3"
 short_desc="Python wrapper to the C++ Parma Polyhedra Library (PPL)"
 maintainer="Gonzalo Tornaría <tornaria@cmat.edu.uy>"
 license="GPL-3.0-or-later"
 homepage="https://gitlab.com/videlec/pplpy"
 changelog="https://gitlab.com/videlec/pplpy/-/raw/master/CHANGES.txt"
 distfiles="${PYPI_SITE}/p/pplpy/pplpy-${version}.tar.gz"
 checksum=500bd0f4ae1a76956fae7fcba77854f5ec3e64fce76803664983763c3f2bd8bd
+
+pre_check() {
+	export PYTHONPATH=$(cd build/lib* && pwd)
```

Comment:
and add python3-pytest to checkdepends
```suggestion
pre_check() {
	PYTHONPATH=$(cd build/lib* && pwd) pytest
```

Reply:
There are a couple extra tests being run by the setup,py test method. Here's the output in the current state:
```
=> python3-pplpy-0.8.7_3: running do_check ...
/builddir/python3-pplpy-0.8.7/setup.py:14: SetuptoolsDeprecationWarning: As setuptools moves its configuration towards `pyproject.toml`,
`setuptools.config.read_configuration` became deprecated.

For the time being, you can use the `setuptools.config.setupcfg` module
to access a backward compatible API, but this module is provisional
and might be removed in the future.

  conf_dict = read_configuration(os.path.join(dir_path, "setup.cfg"))
running test
copying ./tests/testpplpy.pyx -> /tmp/tmp3ane4nvz
creating /tmp/tmp3ane4nvz/__pycache__
copying ./tests/__pycache__/test_constraint.cpython-311-pytest-7.1.3.pyc -> /tmp/tmp3ane4nvz/__pycache__
copying ./tests/__pycache__/test_variable.cpython-311-pytest-7.1.3.pyc -> /tmp/tmp3ane4nvz/__pycache__
copying ./tests/testpplpy2.pyx -> /tmp/tmp3ane4nvz
copying ./tests/setup2.py -> /tmp/tmp3ane4nvz
copying ./tests/test_constraint.py -> /tmp/tmp3ane4nvz
copying ./tests/runtests.py -> /tmp/tmp3ane4nvz
copying ./tests/setup.py -> /tmp/tmp3ane4nvz
copying ./tests/test_variable.py -> /tmp/tmp3ane4nvz
test_creation_invalid (test_variable.TestVariable.test_creation_invalid) ... ok
test_creation_valid (test_variable.TestVariable.test_creation_valid) ... ok

----------------------------------------------------------------------
Ran 2 tests in 0.000s

OK
test_creation_empty (test_constraint.TestConstraint.test_creation_empty) ... ok
test_creation_invalid (test_constraint.TestConstraint.test_creation_invalid) ... ok
test_creation_other (test_constraint.TestConstraint.test_creation_other) ... ok
test_creation_empty (test_constraint.TestConstraint_System.test_creation_empty) ... ok
test_creation_invalid (test_constraint.TestConstraint_System.test_creation_invalid) ... ok
test_creation_other (test_constraint.TestConstraint_System.test_creation_other) ... ok

----------------------------------------------------------------------
Ran 6 tests in 0.001s

OK
Running pplpy doctests
--------------------------------------------------------------------------------
<module 'ppl' from '/builddir/python3-pplpy-0.8.7/build/lib.linux-x86_64-cpython-311/ppl/__init__.py'>
TestResults(failed=0, attempted=34)
--------------------------------------------------------------------------------
<module 'ppl.linear_algebra' from '/builddir/python3-pplpy-0.8.7/build/lib.linux-x86_64-cpython-311/ppl/linear_algebra.cpython-311-x86_64-linux-gnu.so'>
TestResults(failed=0, attempted=402)
--------------------------------------------------------------------------------
<module 'ppl.mip_problem' from '/builddir/python3-pplpy-0.8.7/build/lib.linux-x86_64-cpython-311/ppl/mip_problem.cpython-311-x86_64-linux-gnu.so'>
TestResults(failed=0, attempted=394)
--------------------------------------------------------------------------------
<module 'ppl.polyhedron' from '/builddir/python3-pplpy-0.8.7/build/lib.linux-x86_64-cpython-311/ppl/polyhedron.cpython-311-x86_64-linux-gnu.so'>
TestResults(failed=0, attempted=1204)
--------------------------------------------------------------------------------
<module 'ppl.generator' from '/builddir/python3-pplpy-0.8.7/build/lib.linux-x86_64-cpython-311/ppl/generator.cpython-311-x86_64-linux-gnu.so'>
TestResults(failed=0, attempted=468)
--------------------------------------------------------------------------------
<module 'ppl.constraint' from '/builddir/python3-pplpy-0.8.7/build/lib.linux-x86_64-cpython-311/ppl/constraint.cpython-311-x86_64-linux-gnu.so'>
TestResults(failed=0, attempted=478)
--------------------------------------------------------------------------------
<module 'ppl.congruence' from '/builddir/python3-pplpy-0.8.7/build/lib.linux-x86_64-cpython-311/ppl/congruence.cpython-311-x86_64-linux-gnu.so'>
TestResults(failed=0, attempted=156)
--------------------------------------------------------------------------------
<module 'ppl.bit_arrays' from '/builddir/python3-pplpy-0.8.7/build/lib.linux-x86_64-cpython-311/ppl/bit_arrays.cpython-311-x86_64-linux-gnu.so'>
TestResults(failed=0, attempted=102)
--------------------------------------------------------------------------------
Running unittests
--------------------------------------------------------------------------------
/usr/lib/python3.11/site-packages/Cython/Compiler/Main.py:369: FutureWarning: Cython directive 'language_level' not set, using 2 for now (Py2). This will change in a later release! File: /tmp/tmp3ane4nvz/testpplpy.pyx
  tree = Parsing.p_module(s, pxd, full_module_name)
Compiling testpplpy.pyx because it changed.
[1/1] Cythonizing testpplpy.pyx
running build_ext
building 'testpplpy' extension
creating build
creating build/temp.linux-x86_64-cpython-311
cc -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fstack-clash-protection -D_FORTIFY_SOURCE=2 -mtune=generic -O2 -pipe -g -fstack-clash-protection -D_FORTIFY_SOURCE=2 -mtune=generic -O2 -pipe -g -fstack-clash-protection -D_FORTIFY_SOURCE=2 -mtune=generic -O2 -pipe -fdebug-prefix-map=/builddir/python3-pplpy-0.8.7=. -fPIC -I/builddir/python3-pplpy-0.8.7/build/lib.linux-x86_64-cpython-311/ppl -I/usr/include/python3.11 -c testpplpy.cpp -o build/temp.linux-x86_64-cpython-311/testpplpy.o
creating build/lib.linux-x86_64-cpython-311
g++ -shared -Wl,-z,relro -Wl,-z,now -Wl,--as-needed -Wl,-z,relro -Wl,-z,now -Wl,--as-needed -Wl,-z,relro -Wl,-z,now -Wl,--as-needed -fstack-clash-protection -D_FORTIFY_SOURCE=2 -mtune=generic -O2 -pipe -fdebug-prefix-map=/builddir/python3-pplpy-0.8.7=. build/temp.linux-x86_64-cpython-311/testpplpy.o -L/usr/lib64 -lppl -o build/lib.linux-x86_64-cpython-311/testpplpy.cpython-311-x86_64-linux-gnu.so
copying build/lib.linux-x86_64-cpython-311/testpplpy.cpython-311-x86_64-linux-gnu.so -> 
--------------------------------------------------------------------------------
Cython test 1 OK
--------------------------------------------------------------------------------
Generator_System {ray(1, 0, -1), ray(0, 1, -1), point(0/1, 0/1, 1/1)}
dim = 3
--------------------------------------------------------------------------------
Cython example 1 OK
--------------------------------------------------------------------------------
/usr/lib/python3.11/site-packages/Cython/Compiler/Main.py:369: FutureWarning: Cython directive 'language_level' not set, using 2 for now (Py2). This will change in a later release! File: /tmp/tmp3ane4nvz/testpplpy2.pyx
  tree = Parsing.p_module(s, pxd, full_module_name)
Compiling testpplpy2.pyx because it changed.
[1/1] Cythonizing testpplpy2.pyx
running build_ext
building 'testpplpy2' extension
cc -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fstack-clash-protection -D_FORTIFY_SOURCE=2 -mtune=generic -O2 -pipe -g -fstack-clash-protection -D_FORTIFY_SOURCE=2 -mtune=generic -O2 -pipe -g -fstack-clash-protection -D_FORTIFY_SOURCE=2 -mtune=generic -O2 -pipe -fdebug-prefix-map=/builddir/python3-pplpy-0.8.7=. -fPIC -I/builddir/python3-pplpy-0.8.7/build/lib.linux-x86_64-cpython-311/ppl -I/usr/include/python3.11 -c testpplpy2.cpp -o build/temp.linux-x86_64-cpython-311/testpplpy2.o
g++ -shared -Wl,-z,relro -Wl,-z,now -Wl,--as-needed -Wl,-z,relro -Wl,-z,now -Wl,--as-needed -Wl,-z,relro -Wl,-z,now -Wl,--as-needed -fstack-clash-protection -D_FORTIFY_SOURCE=2 -mtune=generic -O2 -pipe -fdebug-prefix-map=/builddir/python3-pplpy-0.8.7=. build/temp.linux-x86_64-cpython-311/testpplpy2.o -L/usr/lib64 -lppl -o build/lib.linux-x86_64-cpython-311/testpplpy2.cpython-311-x86_64-linux-gnu.so
copying build/lib.linux-x86_64-cpython-311/testpplpy2.cpython-311-x86_64-linux-gnu.so -> 
--------------------------------------------------------------------------------
Cython test 2 OK
--------------------------------------------------------------------------------
Generator_System {ray(1, 0, -1), ray(0, 1, -1), point(0/1, 0/1, 1/1)}
dim = 3
--------------------------------------------------------------------------------
Cython example 2 OK
--------------------------------------------------------------------------------
=> python3-pplpy-0.8.7_3: running pre-install hook: 00-libdir ...
```

The pytest output in comparison:
```
=> python3-pplpy-0.8.7_3: running do_check ...
============================= test session starts ==============================
platform linux -- Python 3.11.0, pytest-7.1.3, pluggy-1.0.0
rootdir: /builddir/python3-pplpy-0.8.7
collected 8 items

tests/test_constraint.py ......                                          [ 75%]
tests/test_variable.py ..                                                [100%]

============================== 8 passed in 0.04s ===============================
=> python3-pplpy-0.8.7_3: running pre-install hook: 00-libdir ...
```

Reply:
hmm, I would still move to pytest, using setup.py is deprecated and someone is going to change it to pytest next time because they won't notice.

Reply:
Do you know why in the python3-module build_style do_check(), the setup.py test section does not have the custom PYTHONPATH like it is in the pytest section? That would resolve having to export PYTHONPATH in some templates I have seen recently.

Reply:
I don't know that, but @ahesford might.

Reply:
Somebody (maybe sgn, but I can't remember for sure) added the path to fix some failing tests. There was no impetus to fix the other alternative.

In fact, given that testing via the setup script is deprecated and most times runs no tests at all, we should just drop it from the build style.

Reply:
Since you are here, any downside of using pytest instead of python3 -m pytest in build styles? (see resolved discussion above for context)

Reply:
I think `python3 -m pytest` solved some issues at the time we added support for pytest in the build style. I'm not sure if it's still relevant. If the build style is changed, we should run through tests in at least a random sample of packages to have confidence that the change isn't causing more harm than good.

Keeping the module form in the pep517 style is probably the right behavior, because it's better to try running tests from the current directory than to fail when the pep517 builder leaves no stages package available for test.

Reply:
Please don't change the way tests are run. The commit claims to "fix tests" but in fact is breaking them in the worst possible way: not running them at all.

Some projects have attempts at using pytest for testing, in the sense that they wrote a couple of silly tests that run via pytest. But then they don't migrate the whole testsuite to pytest. Hence, if you run pytest it might succeed, running a few trivial tests and making you believe tests passed, when in fact the important tests have not run at all.

Reply:
The majority of projects I've seen in our repo silently do nothing with `python3 setup.py test` besides print a deprecation warning. If dropping support for this case highlights more overlooked tests than keeping it does, it's the right thing to do. We can always add back `python3 setup.py test` in manual `do_check` overrides in the templates that still rely on this behavior if the set isn't too large.

Reply:
I meant don't change custom `do_check()` overrides. Also: don't remove a custom `do_check()`. Maybe it's there for a reason. Feel free to ping me if you believe some of my custom `do_check()`s are incorrect or unnecessary.

Here in this thread it is clearly documented that, for this particular package (python3-pplpy), running setup test will run many many more tests. So, don't change it. Same for python3-cypari2 (see #40667).

Reply:
I agree, if custom `do_check` overrides exist, it should be understood that removing it doesn't degrade the test experience before doing so.

Reply:
> The commit claims to "fix tests" but in fact is breaking them in the worst possible way: not running them at all.

Well, my patch was on top of https://github.com/void-linux/void-packages/commit/ecdf957b863134bdfe26208e39d5b877a9ce49f0 which broke the tests - so from my perspective it was indeed a "fix"